<html><body><div><div class="post-text" itemprop="text">

<p><strong>Update</strong>: See "Even Better Solution" below.</p>

<p>This one has me stumped. When I call an HTML page generated by Django REST Framework, it works. When I call it a second, third, fourth time, I get:</p>

<pre><code>[26/Oct/2015 15:14:42]"GET /api/rest/v3/dockets/ HTTP/1.1" 500 92424
Internal Server Error: /api/rest/v3/dockets/
Traceback (most recent call last):
  File "/home/mlissner/.virtualenvs/courtlistener/local/lib/python2.7/site-packages/django/core/handlers/base.py", line 108, in get_response
    response = middleware_method(request)
  File "/home/mlissner/.virtualenvs/courtlistener/local/lib/python2.7/site-packages/django/middleware/cache.py", line 134, in process_request
    response = self.cache.get(cache_key, None)
  File "/home/mlissner/.virtualenvs/courtlistener/local/lib/python2.7/site-packages/django/core/cache/backends/locmem.py", line 54, in get
    return pickle.loads(pickled)
TypeError: __new__() takes exactly 3 arguments (2 given)
</code></pre>

<p>Unlike 99% of the stacktraces I get from Django, this one doesn't mention my code at all, and seems to be only code from Django itself. </p>

<p>I'm using the development server, Django 1.8.7, Django REST Framework 3.2.3, and Python 2.7. </p>

<p>My middleware setting is:</p>

<pre><code>MIDDLEWARE_CLASSES = [
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.contrib.auth.middleware.SessionAuthenticationMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.middleware.cache.UpdateCacheMiddleware',
    'django.middleware.clickjacking.XFrameOptionsMiddleware',
    'corsheaders.middleware.CorsMiddleware',
    'django.middleware.common.CommonMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    'django.middleware.cache.FetchFromCacheMiddleware',
]
</code></pre>

<p>I've only seen this when looking at Django REST Framework pages. Any ideas?</p>

<h2>Things I've Tried</h2>

<ul>
<li>Upgrading Django, djangorestframework and djangorestframework-filters.</li>
<li>Changing my <code>CACHE</code> setting so it uses Redis instead of <code>LocMemCache</code>. I thought this might help because somebody in the comments said changing it to <code>FileBasedCache</code> helped. The change to redis, in itself did not help, though setting it to the <code>DummyCache</code> does.</li>
</ul>

<h2>Solution</h2>

<p><a href="https://django-redis-cache.readthedocs.org/en/latest/advanced_configuration.html#pickle-version" rel="nofollow"><code>django-redis-cache</code></a> allows you to set different versions of pickle so I've tinkered with that, given that <a href="http://bugs.python.org/issue3065" rel="nofollow">one link</a> hinted at the pickle version being related. At first, this seemed to have no effect, but <a href="https://github.com/sebleier/django-redis-cache/issues/108" rel="nofollow">I filed a bug</a> in <code>django-redis</code> ("PICKLE_VERSION doesn't seem to work"), which they quickly fixed. Once that was fixed, I set the PICKLE_VERSION to 1, and the problem was solved. </p>

<p>I've also <a href="https://github.com/tomchristie/django-rest-framework/issues/3725" rel="nofollow">filed a bug in DRF</a> to see if there's a better way to get to the bottom of this. However, I'm not sure if the bug is there, in my code, or in Django itself.</p>

<h2>Even better solution</h2>

<p>I'm the master at workarounds, it seems. But the good news is that this was a <a href="https://github.com/tomchristie/django-rest-framework/issues/3628" rel="nofollow">bug</a> in Django Rest Framework, which has <a href="https://github.com/tomchristie/django-rest-framework/pull/3701" rel="nofollow">been fixed</a>, and will be released in 3.3.2 (hopefully).</p>
    </div>
    </div></body></html>